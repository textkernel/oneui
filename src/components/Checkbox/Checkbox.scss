@import '../../themes/oneui/oneui.scss';

.Checkbox {
    &__box {
        background-color: var(--color-background);
        border: 1px solid var(--color-neutral);
        border-radius: var(--border-radius);
        box-sizing: content-box;
        display: inline-block;
        position: relative;
        top: -1px;
        width: 12px;
        height: 12px;
        transform: scale(1);
        transition: all var(--transition-duration) ease;
        vertical-align: middle;
        margin-right: var(--spacing-2x);

        &--asFlexbox {
            flex-shrink: 0;
        }

        &:before {
            content: '';
            width: 100%;
            height: 100%;
            background: var(--color-brand);
            display: block;
            transform: scale(0);
            opacity: 1;
            border-radius: 50%;
        }
    }

    &__label {
        margin: auto;
        user-select: none;
        cursor: pointer;
        display: block;

        &:hover {
            .Checkbox__box {
                border-color: var(--color-brand);
            }
        }

        &--asFlexbox {
            display: flex;
            align-items: center;
        }
    }

    &__svg {
        position: absolute;
        top: 0;
        left: 0;
        fill: none;
        stroke: var(--color-background);
        stroke-width: 2;
        stroke-linecap: round;
        stroke-linejoin: round;
        transition: all var(--transition-duration) ease;
        transition-delay: 0.1s;
        transform: translate3d(0, 0, 0);
    }

    &__input {
        display: none;

        &--indeterminate {
            + .Checkbox__label {
                .Checkbox__box {
                    background: var(--color-brand);
                    border-color: var(--color-brand);
                }
            }
        }

        &:checked {
            + .Checkbox__label {
                .Checkbox__box {
                    background: var(--color-brand);
                    border-color: var(--color-brand);
                    animation: wave var(--transition-duration) ease;

                    &:before {
                        transform: scale(2.5);
                        opacity: 0;
                        transition: all var(--transition-duration) ease;
                    }
                }
            }
        }
    }

    &__text {
        padding: var(--spacing-normal) 0;
    }

    &--disabled {
        &:hover {
            .Checkbox__label {
                .Checkbox__box {
                    border-color: var(--color-neutral);
                }
            }
        }

        .Checkbox__input {
            + .Checkbox__label {
                cursor: not-allowed;
            }

            &--indeterminate {
                + .Checkbox__label {
                    .Checkbox__box {
                        background: var(--color-neutral);
                        border-color: var(--color-neutral);
                    }
                }
            }

            &:checked {
                + .Checkbox__label {
                    .Checkbox__box {
                        border-color: var(--color-neutral);
                        background: var(--color-neutral);
                    }
                }
            }
        }
    }
}
