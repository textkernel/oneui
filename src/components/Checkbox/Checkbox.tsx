import React, { forwardRef } from 'react';
import { bem } from '../../utils';
import { Text } from '../Text';
import styles from './Checkbox.scss';
import { CHECKBOX_VIEWBOX } from '../../constants';

export interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
    /** A unique id to reference this checkbox */
    id: string;
    /** If the checkbox should be disabled */
    disabled?: boolean;
    /** The label for the checkbox */
    children?: string | React.ReactElement;
    /** Optionally render checkbox in a flexbox */
    asFlexbox?: boolean;
    /** Indeterminate status, show minus sign across checkbox. This property overrides the checked state visually  */
    indeterminate?: boolean;
}

const { block, elem } = bem('Checkbox', styles);

export const Checkbox: React.FC<Props> = forwardRef(
    (
        {
            id,
            children,
            disabled = false,
            asFlexbox = false,
            className,
            style,
            indeterminate = false,
            checked,
            ...rest
        },
        ref
    ) => {
        let text = children;
        if (children && typeof children === 'string') {
            text = (
                <Text {...elem('text')} inline context={disabled ? 'muted' : 'default'}>
                    {children}
                </Text>
            );
        }

        return (
            <div style={style} {...block({ className, asFlexbox })}>
                <input
                    {...rest}
                    {...elem('input', { asFlexbox, indeterminate })}
                    ref={ref}
                    type="checkbox"
                    id={id}
                    disabled={disabled}
                    checked={checked}
                />
                <label {...elem('label', { asFlexbox })} htmlFor={id}>
                    <span {...elem('box', { asFlexbox })}>
                        <svg
                            {...elem('svg', { asFlexbox })}
                            width="12px"
                            height="10px"
                            viewBox={CHECKBOX_VIEWBOX}
                        >
                            {indeterminate && <line x1="2" y1="6" x2="8" y2="6" />}
                            {!indeterminate && (checked === undefined || checked === true) && (
                                <polyline points="1.5 6 3.5 9 8 3" />
                            )}
                        </svg>
                    </span>
                    {text}
                </label>
            </div>
        );
    }
);

Checkbox.displayName = 'Checkbox';
