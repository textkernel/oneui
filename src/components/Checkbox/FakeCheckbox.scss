.FakeCheckbox {
    background-color: var(--color-background-selected-subtlest-default, #E6F2FF);
    border: var(--border-width-thick) solid var(--color-border-input, #808080);
    border-radius: var(--border-width-thick);
    box-sizing: border-box;
    display: inline-block;
    position: relative;
    top: -1px;
    width: var(--font-size-normal);
    height: var(--font-size-normal);
    transition: all var(--transition-duration) ease;
    vertical-align: middle;
    margin-right: var(--space-75);

    &--checked {
        background: var(--color-background-selected-bold-default, #007AFF);
        border-color: var(--color-background-selected-bold-default, #007AFF);
    }

    & svg {
        position: absolute;
        top: var(--border-width-thick);
        left: 0;
        fill: none;
        fill-rule:evenodd;
        clip-rule:evenodd;
        stroke: var(--color-icon-inverse);
        stroke-width: 2;
        stroke-linecap: round;
        stroke-linejoin: round;
    }

}