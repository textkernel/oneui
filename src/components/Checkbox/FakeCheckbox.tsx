import React from 'react';
import { bem } from '../../utils';
import styles from './FakeCheckbox.scss';

export interface Props {
    checked: boolean;
}
const { block } = bem('FakeCheckbox', styles);

export const FakeCheckbox = ({ checked }) => (
    <div {...block({ checked })}>
        <svg width="12px" height="8px" viewBox="0 0 12 9" role="img" aria-hidden="true">
            {checked === undefined || (checked && <polyline points="1.5 5 4 8 10.5 2" />)}
        </svg>
    </div>
);
