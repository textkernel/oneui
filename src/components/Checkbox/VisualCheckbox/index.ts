export { VisualCheckbox, Props as VisualCheckboxProps } from './VisualCheckbox';
