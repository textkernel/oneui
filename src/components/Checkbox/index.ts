export { Checkbox, Props as CheckboxProps } from './Checkbox';
