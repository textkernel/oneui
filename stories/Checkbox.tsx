import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { Checkbox, Text } from '@textkernel/oneui';

storiesOf('Molecules|Checkbox', module)
    .addDecorator(withKnobs)
    .add('Checkbox as controlled component', () => {
        const [checked, setChecked] = React.useState(false);

        const handleChange = (e) => {
            console.log('Checkbox state changed', e);
            setChecked(e.target.checked);
        };

        return (
            <Checkbox
                disabled={boolean('Disabled', false)}
                id={text('Id', 'checkbox-1')}
                onChange={handleChange}
                checked={checked}
                asFlexbox={boolean('Render in flexbox', false)}
            >
                {text('Checkbox label', 'Select me!')}
            </Checkbox>
        );
    })
    .add('Checkbox as uncontrolled component', () => (
        <Checkbox
            disabled={boolean('Disabled', false)}
            id={text('Id', 'checkbox-1')}
            onChange={(e) => console.log('Checkbox state changed', e)}
            asFlexbox={boolean('Render in flexbox', false)}
            defaultChecked
        >
            {text('Checkbox label', 'Select me!')}
        </Checkbox>
    ))
    .add('Checkbox with long label', () => (
        <Checkbox
            disabled={boolean('Disabled', false)}
            id={text('Id', 'checkbox-1')}
            onChange={(e) => console.log('Checkbox state changed', e)}
            asFlexbox={boolean('Render in flexbox', false)}
            defaultChecked
            style={{ width: '150px' }}
        >
            {text(
                'Checkbox label',
                'This is a longer label text. Change flexbox rendering to see how it changes'
            )}
        </Checkbox>
    ))
    .add('Checkbox with not just string as label', () => (
        <Checkbox
            disabled={boolean('Disabled', false)}
            id={text('Id', 'checkbox-1')}
            asFlexbox={boolean('Render in flexbox', false)}
            onChange={(e) => console.log('Checkbox state changed', e)}
        >
            <Text inline style={{ color: 'turquoise' }} className="test-class">
                {text('Checkbox label', 'Select me!')}
            </Text>
        </Checkbox>
    ));
