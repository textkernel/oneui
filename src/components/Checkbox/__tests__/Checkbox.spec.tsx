import * as React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { Checkbox } from '../Checkbox';

describe('<Checkbox> that renders a checkbox', () => {
    const lineLabel = 'line label';
    const polylineLabel = 'polyline label';
    let view: RenderResult;

    it('should render default checkbox correctly', () => {
        view = render(<Checkbox id="c1">Check this out</Checkbox>);

        expect(view.container).toMatchSnapshot();
    });

    it('should render checkbox with complex children', () => {
        view = render(
            <Checkbox id="c1">
                <span>
                    Check this out <span>something else</span>
                </span>
            </Checkbox>
        );

        expect(view.container).toMatchSnapshot();
        expect(
            screen.getByRole('checkbox', { name: 'Check this out something else' })
        ).toBeInTheDocument();
    });

    it('should call onChange function when clicked', async () => {
        const user = userEvent.setup();
        const onChange = jest.fn();
        view = render(
            <Checkbox id="c2" onChange={onChange}>
                Check this out
            </Checkbox>
        );

        await user.click(screen.getByRole('checkbox'));

        expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('should rendered disabled checkbox correctly', () => {
        view = render(
            <Checkbox id="c3" disabled>
                Useless checkbox
            </Checkbox>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('checkbox')).toBeInTheDocument();
        expect(screen.getByRole('checkbox')).toHaveAttribute('disabled');
    });

    describe('svg render', () => {
        const defaultProps = {
            id: 'c4',
            checked: true,
            indeterminate: true,
            onChange: () => {},
            lineLabel,
            polylineLabel,
        };

        beforeEach(() => {
            view = render(<Checkbox {...defaultProps}>Useless checkbox</Checkbox>);
        });

        const rerenderView = (props) => {
            view.rerender(<Checkbox {...defaultProps} {...props} />);
        };

        it('should render the indeterminate svg when checked is false and indeterminate true', () => {
            const newProps = {
                checked: false,
            };

            rerenderView(newProps);

            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole('img', { name: polylineLabel })).not.toBeInTheDocument();
            expect(screen.getByRole('img', { name: lineLabel })).toBeInTheDocument();
        });

        it('should render the checked svg when checked is true and indeterminate false', () => {
            const newProps = {
                id: 'c5',
                indeterminate: false,
            };

            rerenderView(newProps);

            expect(view.container).toMatchSnapshot();
            expect(screen.getByRole('img', { name: polylineLabel })).toBeInTheDocument();
            expect(screen.queryByRole('img', { name: lineLabel })).not.toBeInTheDocument();
        });

        it('should render the indeterminate svg when checked and indeterminate are true', () => {
            const newProps = {
                id: 'c6',
            };

            rerenderView(newProps);

            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole('img', { name: polylineLabel })).not.toBeInTheDocument();
            expect(screen.getByRole('img', { name: lineLabel })).toBeInTheDocument();
        });

        it('should not render any svg when checked and indeterminate are both false', () => {
            const newProps = {
                id: 'c7',
                checked: false,
                indeterminate: false,
            };

            rerenderView(newProps);

            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole(polylineLabel)).not.toBeInTheDocument();
            expect(screen.queryByRole(lineLabel)).not.toBeInTheDocument();
        });
    });
});
