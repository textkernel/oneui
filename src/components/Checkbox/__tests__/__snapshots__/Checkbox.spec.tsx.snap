// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> that renders a checkbox should render checkbox with complex children 1`] = `
<div>
  <div>
    <input
      class="Checkbox__input"
      id="c1"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c1"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <polyline
            points="1.5 6 3.5 9 8 3"
          />
        </svg>
      </span>
      <span>
        Check this out 
        <span>
          something else
        </span>
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox should render default checkbox correctly 1`] = `
<div>
  <div>
    <input
      class="Checkbox__input"
      id="c1"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c1"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <polyline
            points="1.5 6 3.5 9 8 3"
          />
        </svg>
      </span>
      <span
        class="Checkbox__text"
      >
        Check this out
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox should rendered disabled checkbox correctly 1`] = `
<div>
  <div>
    <input
      class="Checkbox__input"
      disabled=""
      id="c3"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c3"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <polyline
            points="1.5 6 3.5 9 8 3"
          />
        </svg>
      </span>
      <span
        class="Text--context_neutral Checkbox__text"
      >
        Useless checkbox
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox svg render should not render any svg when checked and indeterminate are both false 1`] = `
<div>
  <div>
    <input
      checked=""
      class="Checkbox__input"
      id="c7"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c7"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          aria-label="polyline label"
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        />
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox svg render should render the checked svg when checked is true and indeterminate false 1`] = `
<div>
  <div>
    <input
      checked=""
      class="Checkbox__input"
      id="c5"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c5"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          aria-label="polyline label"
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <polyline
            points="1.5 6 3.5 9 8 3"
            role="polyline label"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox svg render should render the indeterminate svg when checked and indeterminate are true 1`] = `
<div>
  <div>
    <input
      checked=""
      class="Checkbox__input Checkbox__input--indeterminate"
      id="c6"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c6"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          aria-label="line label"
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <line
            role="line label"
            x1="2"
            x2="8"
            y1="6"
            y2="6"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;

exports[`<Checkbox> that renders a checkbox svg render should render the indeterminate svg when checked is false and indeterminate true 1`] = `
<div>
  <div>
    <input
      checked=""
      class="Checkbox__input Checkbox__input--indeterminate"
      id="c4"
      type="checkbox"
    />
    <label
      class="Checkbox__label"
      for="c4"
    >
      <span
        class="Checkbox__box"
      >
        <svg
          aria-label="line label"
          class="Checkbox__svg"
          height="10px"
          role="img"
          viewBox="0,0,10,10"
          width="12px"
        >
          <line
            role="line label"
            x1="2"
            x2="8"
            y1="6"
            y2="6"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;
